### ISSUE TITLE
Notification message key overwritten

### ISSUE THREAD

**Issue by thomasconner** [phonegap/phonegap-plugin-push#1710]
At [Kinvey](http://www.kinvey.com), push notifications are sent to android devices with the following payload `{ message: "{"message": "You have a new message", "from": "Kinvey", "subject": "Welcome to BL"}"}`.

### Expected Behavior
Expected the notification to be shown with the nested message key value `"You have a new message"`.

### Actual Behavior
A notification is shown with the message key value set to `"{"message": "You have a new message", "from": "Kinvey", "subject": "Welcome to BL"}"`.

### Reproduce Scenario (including but not limited to)
None

#### Steps to Reproduce 
Send a push notification use the [Kinvey Console](http://www.console.kinvey.com) to an Android PhoneGap application.

#### Platform and Version (eg. Android 5.0 or iOS 9.2.1)
Android 7.0

#### (Android) What device vendor (e.g. Samsung, HTC, Sony...)
Nexus

#### Cordova CLI version and cordova platform version
cordova --version: 7.0.0
cordova platform version android: 6.2.3

#### Plugin version
1.10.3

#### Sample Push Data Payload
`{ message: "{"message": "You have a new message", "from": "Kinvey", "subject": "Welcome to BL"}"}`

#### Sample Code that illustrates the problem
None

#### Logs taken while reproducing problem
None

**Comment by thomasconner** [comment]
If I comment out lines 258 to 260 in https://github.com/phonegap/phonegap-plugin-push/blob/master/src/android/com/adobe/phonegap/push/GCMIntentService.java#L258 then the push notification is shown correctly.

What is the purpose of this code?

**Comment by macdonst** [comment]
@thomasconner it's to make the code later on much cleaner so I don't have to do a bunch of if's where I have to account for a bunch of different keys that all are just the push notification's message/body/etc. The old PushPlugin used `message` as a key which is why I kept it around but it should be `body`.

**Comment by thomasconner** [comment]
@macdonst I wouldn't mind to help out and submit a PR but I am not sure which key should be `body` instead of `message`.

**Comment by thomasconner** [comment]
I thought a few screenshots of the issue would help. I sent a notification that contained the payload `{ message: "{"message": "You have a new message", "from": "Kinvey", "subject": "Welcome to BL"}"}`.

This screenshot shows the message incorrectly.
![screenshot2](https://cloud.githubusercontent.com/assets/950052/25868049/92e3d0b8-34c9-11e7-9ed4-76b1ecce66f1.png)

This screenshot shows the message correctly. I have commented out lines 258 to 260 https://github.com/phonegap/phonegap-plugin-push/blob/master/src/android/com/adobe/phonegap/push/GCMIntentService.java#L258 for this screenshot.
![screenshot1](https://cloud.githubusercontent.com/assets/950052/25868051/957445ba-34c9-11e7-9a8d-95a23a22864a.png)

**Comment by macdonst** [comment]
@thomasconner yeah, I get what is going on. If you look at this line:

https://github.com/phonegap/phonegap-plugin-push/blob/master/src/android/com/adobe/phonegap/push/GCMIntentService.java#L221

You'll see if the key is `message` it will use that as the main text in the notification. However, we should be checking to see if the value of `message` is a json object. If it is then we should loop through it's keys instead of assuming the whole thing is a string.

**Comment by fredgalvao** [comment]
Related to #1713?

**Comment by thomasconner** [comment]
@macdonst If you look at [line 216](https://github.com/phonegap/phonegap-plugin-push/blob/master/src/android/com/adobe/phonegap/push/GCMIntentService.java#L216) there is already a check to see if the value of `message` is a json object. Once the `while` loop completes, `newExtras` correctly contains the parsed `message` value. However on [line 260](https://github.com/phonegap/phonegap-plugin-push/blob/master/src/android/com/adobe/phonegap/push/GCMIntentService.java#L260) it replaces the value of `message` in `newExtras` with the json object value in `extras`.

I think the fix should be to flip the arguments for `replaceKey` on [line 260](https://github.com/phonegap/phonegap-plugin-push/blob/master/src/android/com/adobe/phonegap/push/GCMIntentService.java#L260).

```javascript
replaceKey(context, key, newKey, newExtras, extras); // flipped newExtras and extras
```

**Comment by thomasconner** [comment]
This is related to PR https://github.com/phonegap/phonegap-plugin-push/pull/1666

**Comment by macdonst** [comment]
@thomasconner can you send me the full push payload? I'm seeing some oddities to be sure but I want to make sure I test with the correct payload. For instance here is a sample payload I'm using:

```
{
       "registration_ids": "fake_id",
        "data" : {
	    	"title": "Large Circular Icon",
	    	"body": "Loaded from URL",
        	"image": "https://pbs.twimg.com/profile_images/837060031895896065/VHIQ4oUf_400x400.jpg",
        	"image-type": "circle",
        	"no-cache": "0"
        }
```

Where does your `message` property fit?

**Comment by thomasconner** [comment]
@macdonst See the payload below. I think this is the format it is sent.

```
{
    "registration_ids": "fake_id",
    "data" : {
        "message": "{"message": "You have a new message", "from": "Kinvey", "subject": "Welcome to BL"}"
    }
}
```

**Comment by macdonst** [comment]
@thomasconner Thanks! Sorry for the delay on this. I've been traveling all around the EU this month giving talks. So many planes.

**Comment by macdonst** [comment]
@thomasconner I merged PR #1666 which fixes the issue where the message property is double normalized then did a small fix to pull to use the subject you sent ad the title of the push message as well. Install the latest from github master and let me know if that resolves all your issues. If it does I will do a bug fix release almost immediately.

**Comment by thomasconner** [comment]
@macdonst Thanks for merging the PR. This did resolve the issue when the `message` key is a JSON string value. The push notification is now showing correctly.

However, there is one more case that is now missing and not working. Sometimes a push notification is sent with a much simpler payload like the following below.

```
{
    "registration_ids": "fake_id",
    "data" : {
        "message": "Welcome to BL"
    }
}
```

For this case the push notification is not shown because the message key is removed from the `extras`. On [line 216|https://github.com/phonegap/phonegap-plugin-push/blob/master/src/android/com/adobe/phonegap/push/GCMIntentService.java#L216] we need a matching else statement to handle the case where the string is not a JSON serialized object.

```javascript
// line 216
if (json instanceof String && ((String) json).startsWith("{")) {
    // ...
} else {
    String newKey = normalizeKey(key, messageKey, titleKey);
    Log.d(LOG_TAG, "replace key " + key + " with " + newKey);
    replaceKey(context, key, newKey, extras, newExtras);
}
```

**Comment by thomasconner** [comment]
@macdonst Let me know if you would like me to submit a PR for this change.

**Comment by macdonst** [comment]
@thomasconner I just pushed up the change after doing a quick bit of testing. Seems good to me. Let me know if it satisfies things on your end.

**Comment by thomasconner** [comment]
@macdonst Everything worked great after that last change. Please if you don't mind doing a release as soon as you are able to this would really help. Thanks again for all the help!

**Comment by macdonst** [comment]
@thomasconner published 1.10.4. Hmmm…I'm doing a workshop on push tomorrow I hope I didn't screw anything up.
